New instrument view single plotter in GUI and detector table

GuiMacielPereira opened the pull request:
### Description of work
This work improves the GUI of the instrument view by showing only a single plotter instead of two.
It also simplifies the model by using `CreateDetectorTable` to get all the necessary arrays instead of using `ws.componentInfo()` which is slower and more cumbersome.

<!-- Please provide an outline and reasoning for the work.
If there is no linked issue provide context.
-->

Closes #[39867](https://github.com/mantidproject/mantid/issues/39867) and #[39745](https://github.com/mantidproject/mantid/issues/39745) <!-- One line per closed issue. -->

<!-- If issue raised by user. Do not leak email addresses.
**Report to:** [user name]
-->

### To test:

<!-- Include sufficient instructions for someone unfamiliar with the application to test.
Ok to refer back to instructions in the issue.
-->
Code review and functional testing:

- Make sure to tick/untick box for rectangular section 3D view
- Change projection type and make sure that colorbar and counts are the same
- Change projection type and check that picked detectors are the same
- Try several instruments

<!-- REMEMBER:
- Add labels, milestones, etc.
- Ensure the base of this PR is correct (e.g. release-next or main)
- Add release notes in separate file as per ([guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)), or justify their absence:
  *This does not require release notes* because <fill in an explanation of why>
-->

<!--  GATEKEEPER: When squashing, remove the section from HERE...  -->
---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

GuiMacielPereira committed:
Change view to have a single plotter

- Removed duplicated code from view and presenter
- Deal with tof_limits and counts_limits in model (as they are a state)

GuiMacielPereira committed:
Use detector table to initialise attributes of model

Cleaned up model initialisations by using the detector table and
explicitly set getters with the property decorator

GuiMacielPereira committed:
Fix issue with rectangle select

- Disabled box for projections since it's not yet a feature
- Fixed bug where clicking 'r' to start selection and then unticking
rectangle select box caused plotter to be stuck with wrong interactions

GuiMacielPereira committed:
Update unit tests

jclarkeSTFC reviewed (dismissed):
I tested out the IV with workspaces from a few instruments (including WISH), these changes make the IV much faster to load. I also much prefer having the line plot where it is in this PR, because it's important info and deserves more space than a box in the corner.

RichardWaiteSTFC reviewed (changes requested):
Just seen a few small things, btu glad the detector table has provided a speed up!

GuiMacielPereira committed:
Change min,max tof limits to consider ragged workspaces

Ragged workspaces will have tof limits that might be excluded from
ws.extractX(). To be correct, need to loop over spectra one by one.

GuiMacielPereira commented on qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py:
```diff
@@ -54,8 +54,15 @@ def setup(self):
+        last_values = np.zeros_like(first_values)
+        for i in range(self._workspace.getNumberHistograms()):
+            data_x = self._workspace.dataX(i)
+            first_values[i] = data_x[0]
```
Yeah, my first reasoning was that calling min and max multiple times is slower than calling it a single time on a big array

GuiMacielPereira committed:
Improve tof_limits and add unit test

Use different forms of calculating the tof limits depending on the kind
of workspace.

GuiMacielPereira commented on qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py:
```diff
@@ -54,8 +54,15 @@ def setup(self):
+        # Get min and max tof values, loop required because of ragged workspaces
+        first_values = np.zeros(self._workspace.getNumberHistograms())
+        last_values = np.zeros_like(first_values)
+        for i in range(self._workspace.getNumberHistograms()):
```
If `self._workspace.extratX()` returns a reference then assigning it to the variable will not copy an entire array, but will rather be a view of the array (like a pointer in C++), so I still assigned it to a variable because makes the code a bit easier to read.

GuiMacielPereira commented on qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py:
```diff
@@ -28,96 +28,136 @@ def __init__(self, workspace: Workspace2D):
+        # Get min and max tof values
+        if self._workspace.isRaggedWorkspace():
+            first_last = [self._workspace.readX(i)[[0, -1]] for i in range(self._workspace.getNumberHistograms())]
+            self._tof_limits = (np.min(first_last), np.max(first_last))
```
Yeah, I was just avoiding creating an array for that list comprehension but since we need to call numpy methods on it, it's already going to be created by numpy regardless, so I don't think there was a performance improvement. Now that I think about it, maybe calling min and max on the list is slower because it probably creates the array anew at each call
